File: 1302-deepest-leaves-sum/1302-deepest-leaves-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        lsum, maxDepth = 0, 0
        def dfs(root, depth):
            if not root: return depth - 1
            if not root.left and not root.right:
                return depth
            left = dfs(root.left, depth + 1)
            right =dfs(root.right, depth + 1)
            return max(left, right)
        
        maxDepth= dfs(root, 0)
        
        def dfs2(root, depth):
            nonlocal maxDepth,lsum
            if depth == maxDepth:
                lsum += root.val
            if root.left:
                dfs2(root.left, depth + 1)
            if root.right:
                dfs2(root.right, depth + 1)
        dfs2(root, 0)
        return lsum

File: 1302-deepest-leaves-sum/1302-deepest-leaves-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    int maxDepth = 0;
    public int deepestLeavesSum(TreeNode root) {
        dfs(root, 0);
        return res;
    }
    public int dfs(TreeNode root, int depth) {
        if(root == null) return 0;
        if(root.left == null && root.right == null) { //leave
            if(depth > maxDepth) {
                maxDepth = depth;
                res = 0;
                res += root.val;
            } else if(depth == maxDepth) {
                res += root.val;
            }
            
        }
        return dfs(root.left, depth + 1) + dfs(root.right, depth + 1);
    }
}